internal/pkg/util/fs/helper.go

```diff
@@ -176,23 +176,33 @@ func MakeTmpFile(basedir, pattern string, mode os.FileMode) (*os.File, error) {
 	return f, nil
 }
 
-// CopyFile copies file to the provided location making sure the resulting
-// file has permission bits set to the mode.
-func CopyFile(from, to string, mode os.FileMode) (err error) {
-	dstFile, err := MakeTmpFile(filepath.Dir(to), "", mode)
+// FileExists simply checks if a path exists.
+func FileExists(path string) (bool, error) {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// CopyFile copies file to the provided location. To honor umask
+// correctly, the to file must not exist.
+func CopyFile(from, to string, mode os.FileMode) error {
+	exist, err := FileExists(to)
 	if err != nil {
-		return fmt.Errorf("could not create temp file: %v", err)
+		return err
+	}
+	if exist {
+		return fmt.Errorf("file %s already exists", to)
 	}
 
-	defer func() {
-		dstFile.Close()
-		if err == nil {
-			if e := os.Rename(dstFile.Name(), to); e != nil {
-				err = fmt.Errorf("could not rename temp file: %v", err)
-			}
-		}
-		os.Remove(dstFile.Name())
-	}()
+	dstFile, err := os.OpenFile(to, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, mode)
```

Comment:
This is wrong, if `io.Copy` fails you will have corrupted file here. 
The creation of temp file and defer was not added to handle this particular case, please return that code back.
See https://github.com/sylabs/singularity/pull/4179#discussion_r313509655

Reply:
This is ~wrong~ **correct**, as it will honor umask. The issue with the tmp file, is that the `os.Chmod()` will override umask. There is a way to have have them both, which we can talk about.

Reply:
What do you mean by `The issue with the tmp file, is that the os.Chmod() will override umask.`? Do . you mean file permissions here (because that is what `mode` is)? The same permissions are applied during temp file creation,  and afaik `os.Rename` which is `mv` for Linux does not change them.

Reply:
os.Rename() is not the issue. The issue is that when doing the `os.Chmod()`, that will override the umask. So the solution is to create the tmp file with `os.OpenFile(...)`.